### ISSUE TITLE
Extract Vips image and save it to numpy array same as with PIL?

### ISSUE THREAD

**Issue by alfredox10** [jcupitt/libvips#318]
I have looked through some of the documentation available [Here][1] for Vips but have not found an answer yet.

I want to put the image into a numpy 3D array similar to how PIL images automatically get processed this way:

    In[1]:  import numpy
    In[2]:  from PIL import Image
    In[3]:  image = Image.open('43.jpg')
    In[4]:  image
    Out[4]: <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=3216x2136 at 0x7F0C8D8B9950>
    In[5]:  imgArray2 = numpy.asarray(image, dtype=numpy.float32)
    In[6]:  imgArray2.shape
    Out[6]: (2136, 3216, 3)

.

This is what I have for Vips so far...

    In[1]:  import numpy
    In[2]:  from gi.repository import Vips
    In[3]:  image = Vips.Image.new_from_file('43.jpg')
    In[4]:  image
    Out[4]: <Image object at 0x7f0c9a66c5f0 (VipsImage at 0x338a190)>
    In[5]:  imgArray2 = numpy.asarray(image, dtype=numpy.float32)
    Out[5]: ValueError: setting an array element with a sequence.

So I get that error at the end because I'm not pulling the data in the right format from the Vips Image object. 

  [1]: http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/using-from-python.html "here"

**Comment by jcupitt** [comment]
You need `.write_to_memory()` and `.new_from_memory()`. The C docs are here:

http://www.vips.ecs.soton.ac.uk/supported/8.0/doc/html/libvips/VipsImage.html#vips-image-write-to-memory

http://www.vips.ecs.soton.ac.uk/supported/8.0/doc/html/libvips/VipsImage.html#vips-image-new-from-memory

You use them in Python like this:

```
$ python
Python 2.7.9 (default, Apr  2 2015, 15:33:21)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from gi.repository import Vips
>>> x = Vips.Image.new_from_file("/home/john/pics/k2.jpg")
>>> y = x.write_to_memory()
>>> type(y)
<type 'str'>
>>> z = Vips.Image.new_from_memory(y, x.width, x.height, x.bands, Vips.BandFormat.UCHAR)
>>> x.avg()
102.79218031609196
>>> z.avg()
102.79218031609196
```

So that's sending the JPEG file off to a string, then creating a new image back from the string again. The string is just the sequence of bytes which make up the image values, so for an 8-bit RGB image like this, it will start with:

   R1 G1 B1

where R1 is the value of the red channel for the left-left pixel in the image. Pixels are stored as a series of scanlines, top to bottom.

**Comment by jcupitt** [comment]
Did this work? Please reopen if it didn't.

**Comment by alfredox10** [comment]
That didn't work because I was using numpy float arrays so it gave me errors, even when I changed the Vips.BandFormat to float. This is what worked:

from gi.repository import Vips
from PIL import Image
import numpy as np

with open(path_or_url) as image_file:

  start_pillow = time.time()
  pillow_img = np.asarray(Image.open(image_file))
  print('Pillow Time:', time.time()-start_pillow)
  print('original shape', pillow_img.shape)

  start_vips = time.time()
  img = Vips.Image.new_from_file(path_or_url)
  print('Image bit depth', img.Bbits)
  mem_img = img.write_to_memory()

  # Note that my specific image was 8 bit
  np_3d = np.fromstring(mem_img, dtype=np.uint8).reshape(img.width, img.height, 3)

  print('Vips Time:', time.time()-start_vips)
  print('final shape', np_3d.shape)

  # Just to verify we have the same result
  print('Sum of the Differences:', np.sum(np_3d-pillow_img))

**Comment by blipk** [comment]
I tried the above code and it seems the reshape isn't working correctly, or I'm not understanding numpy arrays correctly

```
    image = pyvips.Image.new_from_file(address+'_page.png', access='sequential')
    image = image.crop(720, 300, 500, 400)

    img_mem = image.write_to_memory()
    np_3d = np.frombuffer(img_mem, dtype=np.uint8).reshape(image.width, image.height, 4)

    print(np_3d.shape)
    print(image.getpoint(243, 179))
    print(np_3d[243, 179])
```
Outputs:
```
(500, 400, 4)
[72.0, 176.0, 221.0, 255.0]
[218 216 214 255]
```

The pixels at those positions should be the same? Am I doing something wrong?

**Comment by jcupitt** [comment]
There's a chapter in the pyvips docs about this:

https://libvips.github.io/pyvips/intro.html#numpy-and-pil

tldr: use `np.ndarray()`.

Here's a sample program:

```python
#!/usr/bin/python3

import sys
import os
import psutil

import pyvips
import numpy as np
import cv2

format_to_dtype = {
    'uchar': np.uint8,
    'char': np.int8,
    'ushort': np.uint16,
    'short': np.int16,
    'uint': np.uint32,
    'int': np.int32,
    'float': np.float32,
    'double': np.float64,
    'complex': np.complex64,
    'dpcomplex': np.complex128,
}

dtype_to_format = {
    'uint8': 'uchar',
    'int8': 'char',
    'uint16': 'ushort',
    'int16': 'short',
    'uint32': 'uint',
    'int32': 'int',
    'float32': 'float',
    'float64': 'double',
    'complex64': 'complex',
    'complex128': 'dpcomplex',
}

process = psutil.Process(os.getpid())

def pyvips_to_numpy(image):
    return np.ndarray(buffer=image.write_to_memory(),
                      dtype=format_to_dtype[image.format],
                      shape=[image.height, image.width, image.bands])

def numpy_to_pyvips(npimage):
    height, width, bands = npimage.shape
    linear = npimage.reshape(width * height * bands)
    return pyvips.Image.new_from_memory(linear.data, width, height, bands,
                                        dtype_to_format[str(npimage.dtype)])

def process_size():
    print(f"process size: {process.memory_info().rss / (1024 * 1024):.2f} MB")

print("loading with numpy ...")
npimage = cv2.imread(sys.argv[1])
process_size()

print("converting to pyvips ...")
image = numpy_to_pyvips(npimage)
process_size()

print("rotating with pyvips and creating new numpy image ...")
image = image.rotate(346.2)
npimage2 = pyvips_to_numpy(image)
process_size()

print("writing to output ...")
cv2.imwrite(sys.argv[2], npimage2)
process_size()
```

Unless you're doing scientific imaging, your image files are probably just uint8, so you can make `pyvips_to_numpy` just:

```
np.ndarray(buffer=image.write_to_memory(),
                      dtype=np.uint8,
                      shape=[image.height, image.width, image.bands])
```
